<launch>

<arg name="map_size_x" default="600.0"/>
<arg name="map_size_y" default="100.0"/>
<arg name="map_size_z" default="30.0"/>
<arg name="map_frame_name"  default="world"/>

<arg name="start_x" default=" -260.0"/>
<arg name="start_y" default=" 0.0"/>
<arg name="start_z" default=" 16.0"/>

<arg name="sensing_horizon"  default="50.0" />
<arg name="sensing_rate"     default="20.0"/>
<arg name="estimation_rate"  default="20.0"/>

  <!-- Trajectory Optimization -->
  <node pkg="trajectory_generator" type="trajectory_generator_node" name="trajectory_generator_node" output="screen" required = "true">
      <remap from="~waypoints"        to="/waypoint_generator/waypoints"/>
      <remap from="~local_pointcloud" to="/pcl_render_node/local_pointcloud"/> 
      <remap from="~odom"             to="/visual_slam/odom"/>
      
      <param name="planning/planning_alg"       value="1"    />
      <param name="planning/if_2d_search"       value="1"    />
      <param name="planning/h_reference"        value="15.0" />
      <param name="planning/vel"                value="3.0"  />
      <param name="planning/acc"                value="2.0"  />
      <param name="planning/dev_order"          value="4"    />
      <param name="planning/min_order"          value="3"    />
      <param name="vis/vis_traj_width"          value="0.5" />
      <param name="map/margin"                  value="0.0" />
      <param name="map/resolution"              value="2.0" />
      <param name="map/x_size"                  value="$(arg map_size_x)"/>
      <param name="map/y_size"                  value="$(arg map_size_y)"/>
      <param name="map/z_size"                  value="$(arg map_size_z)"/>
      <param name="replanning/thresh_replan"    value="5.0" type="double"/>
      <param name="replanning/thresh_no_replan" value="15.0" type="double"/>
      <param name="path/resolution"             value="0.2"/>
      <param name="traj_server/time_forward"    value="0.3"/>
      
      <param name="optimization/lambda_smooth" value="2.0" type="double"/>
      <param name="optimization/lambda_collision" value="0.76" type="double"/>
      <param name="optimization/lambda_feasibility" value="0.04" type="double"/>
      <param name="optimization/lambda_curvature" value="1" type="double"/>
      <param name="optimization/lambda_fitness" value="0.00" type="double"/>
      <param name="optimization/dist0" value="8.0" type="double"/>
      <param name="optimization/max_vel" value="12.0" type="double"/>
      <param name="optimization/min_vel" value="8.0" type="double"/>
      <param name="optimization/max_K" value="0.00" type="double"/>    <!--<param name="map/random_seed" value="40.0"/>-->     
      <param name="optimization/max_acc" value="12.0" type="double"/>
      <param name="optimization/order_" value="3" type="int"/>
      <param name="optimization/control_points_distance" value="4.0" type="double"/>

      <param name="hovering_speed" value="6.0" type="double"/>
      <param name="hovering_radius" value="20.0" type="double"/>

      <param name="ego_planner_manager/max_vel"                 value="3.0"   type="double"/>
      <param name="ego_planner_manager/max_acc"                 value="2.0"   type="double"/>
      <param name="ego_planner_manager/max_jerk"                value="4.0"   type="double"/>
      <param name="ego_planner_manager/control_points_distance" value="4.0"   type="double"/>
      <param name="ego_planner_manager/feasibility_tolerance"   value="0.05"  type="double"/>
      <param name="ego_planner_manager/planning_horizon"        value="$(arg sensing_horizon)" type="double"/>
      <param name="ego_planner_grid_map/local_update_range_x"   value="25" /> 
      <param name="ego_planner_grid_map/local_update_range_y"   value="25" /> 
      <param name="ego_planner_grid_map/local_update_range_z"   value="25" /> 
      <param name="ego_planner_grid_map/obstacles_inflation"    value="1.0"/>
      <param name="ego_planner_grid_map/obstacles_inflation"    value="1.0"/> 


  </node>

  <!-- Trajectory Command server -->
  <node pkg="trajectory_generator" type="trajectory_server" name="trajectory_server" output="screen" launch-prefix="">
        <remap from="~/odometry" to="/visual_slam/odom"/>
        <remap from="~/trajectory" to="/trajectory_generator_node/trajectory"/>
  </node>

<!-- Random Map Generator -->
<node pkg ="map_generator" name ="random_complex" type ="random_complex" output = "screen">    
    <remap from="~odometry"       to="/visual_slam/odom"/>

    <param name="init_state_x"        value="$(arg start_x)"/>
    <param name="init_state_y"        value="$(arg start_y)"/>

    <param name="map/x_size"          value="$(arg map_size_x)" />
    <param name="map/y_size"          value="$(arg map_size_y)" />
    <param name="map/z_size"          value="$(arg map_size_z)" />
    <param name="map_frame_name"      value="$(arg map_frame_name)"/>

    <param name="map/circle_num" value="0"/>        
    <param name="map/obs_num"    value="66"/>        
    <param name="map/resolution" value="1"/> 
    <param name="map/random_seed" value="40.0"/>
 

    <param name="ObstacleShape/lower_rad" value="4"/>
    <param name="ObstacleShape/upper_rad" value="8"/>
    <param name="ObstacleShape/lower_hei" value="10.0"/>
    <param name="ObstacleShape/upper_hei" value="20.0"/>
    <param name="CircleShape/lower_circle_rad"   value="0.6"/>        
    <param name="CircleShape/upper_circle_rad"   value="3.5"/>        
    
    <param name="sensing/rate"   value="2.0"/>       
</node>

<!-- Map Info Render -->
<node pkg="map_render" type="pcl_render_node" name="pcl_render_node" output="screen">
    <param name="sensing_horizon"   value="$(arg sensing_horizon)" />
    <param name="sensing_rate"      value="$(arg sensing_rate)"/>
    <param name="estimation_rate"   value="$(arg estimation_rate)"/>
    <param name="map/x_size"        value="$(arg map_size_x)"/>
    <param name="map/y_size"        value="$(arg map_size_y)"/>
    <param name="map/z_size"        value="$(arg map_size_z)"/>
    <param name="map_frame_name"    value="$(arg map_frame_name)"/>

    <remap from="~global_map"       to="/random_complex/global_map"/>
    <remap from="~global_ground"    to="/random_complex/global_ground"/>
    <remap from="~local_map"        to="/click_obstacle_generation/click_new_obs"/>
    <remap from="~odometry"         to="/visual_slam/odom"/>
</node>
  <![CDATA[
   <!-- Quadrotor Simulator -->
   <node pkg="so3_quadrotor_simulator" type="quadrotor_simulator_so3" name="quadrotor_simulator_so3" output="screen">
     <param name="rate/odom" value="100.0"/>

     <param name="simulator/init_state_x" value="$(arg start_x)"/>
     <param name="simulator/init_state_y" value="$(arg start_y)"/>
     <param name="simulator/init_state_z" value="$(arg start_z)"/>
     <remap from="~odom" to="/visual_slam/odom"/>
     <remap from="~cmd" to="so3_cmd"/>
     <remap from="~imu" to="sim/imu"/>     
   </node>
  ]]>
   <!-- Fixed-wing Controller -->
    <rosparam command="load" file="$(find dynamics)/param/plane_params.yaml" />
    <rosparam command="load" file="$(find autopilot)/param/control_params.yaml" />
    
    <node pkg="dynamics" type="dynamics" name="dynamics" output="screen" />

    <node pkg="estimator" type="ekf" name="ekf" output="screen" />
      <param name="odom_rate" value="100.0"/>
      <remap from="odom" to="/visual_slam/odom"/>

    <node pkg="autopilot" type="autopilot" name="autopilot" output="screen" />

    <node pkg="dynamics" type="windsim" name="wind" output="screen" />
      <remap from="odom"            to="/visual_slam/odom"/>
    <![CDATA[
   <!-- Quadrotor Controller -->  
   <node pkg="nodelet" type="nodelet" args="standalone so3_control/SO3ControlNodelet" name="so3_control" required="true" output="screen">
     <remap from="~odom"          to="/visual_slam/odom"/>
     <remap from="~position_cmd"  to="/traj_server/position_command"/>
     <remap from="~motors"        to="motors"/>
     <remap from="~corrections"   to="corrections"/>
     <remap from="~so3_cmd"       to="so3_cmd"/>
     <remap from="~imu"           to="sim/imu"/>

   <rosparam file="$(find so3_control)/config/gains_hummingbird.yaml"/>
     <rosparam file="$(find so3_control)/config/corrections_hummingbird.yaml"/>
     <param name="mass"                   value="0.98"/>
     <param name="use_angle_corrections " value="false"/>
     <param name="use_external_yaw "      value="false"/>
     <param name="gains/rot/z"            value="1.0"/>    
     <param name="gains/ang/z"            value="0.1"/>        
   </node> 
    ]]>
   <!-- Odom Visualization -->
   <node pkg="odom_visualization" name="odom_visualization" type="odom_visualization" output="screen">
     <remap from="~odom"            to="/visual_slam/odom"/>
     <param name="color/a"          value="0.8"/>    
     <param name="color/r"          value="1.0"/>        
     <param name="color/g"          value="0.50"/>        
     <param name="color/b"          value="0.25"/>       
     <param name="covariance_scale" value="100.0"/>
     <param name="drone_scale"      value="0.5"/> 
     <param name="cross_config"      value="true"/>       
   </node>

  <!-- Set the goal -->
  <node pkg="waypoint_generator" name="waypoint_generator" type="waypoint_generator" output="screen">
      <remap from="~goal"           to="/goal"/>
      <param name="waypoint_type"   value="manual-lonely-waypoint"/>    
  </node>

  <!-- Start rviz -->
  <node name="rviz" pkg="rviz" type="rviz" output="screen"
       args="-d $(find trajectory_generator)/launch/rviz_config/demo.rviz"> 
  </node>

</launch>
